x-pack/platform/plugins/shared/fleet/server/tasks/sync_integrations/sync_integrations_on_remote.ts

```diff
@@ -180,3 +215,219 @@ export const syncIntegrationsOnRemote = async (
+    const { integrations: ccrIntegrations, custom_assets: ccrCustomAssets } = ccrIndex;
+
+    // find integrations installed on remote
+    const installedIntegrations = await getPackageSavedObjects(savedObjectsClient);
+
+    if (!installedIntegrations && ccrIntegrations.length > 0) {
+      return {
+        integrations: [],
+        error: `No integrations installed on remote`,
+      };
+    }
+
+    const installedIntegrationsByName = (installedIntegrations?.saved_objects || []).reduce(
+      (acc, integration) => {
+        if (integration?.id) {
+          acc[integration.id] = integration;
+        }
+        return acc;
+      },
+      {} as Record<string, SavedObjectsFindResult<Installation>>
+    );
+    const customAssetsStatus = await compareCustomAssets(esClient, logger, ccrCustomAssets);
+    const integrationsStatus = compareIntegrations(ccrIntegrations, installedIntegrationsByName);
+    const result = {
+      ...integrationsStatus,
+      ...(customAssetsStatus && { custom_assets: customAssetsStatus }),
+    };
+
+    return result;
+  } catch (err) {
+    logger.error('error', err.message);
+    throw err;
+  }
+};
+
+const compareIntegrations = (
+  ccrIntegrations: IntegrationsData[],
+  installedIntegrationsByName: Record<string, SavedObjectsFindResult<Installation>>
+) => {
+  const integrationsStatus: RemoteSyncedIntegrationsStatus[] | undefined = ccrIntegrations?.map(
+    (ccrIntegration) => {
+      const localIntegrationSO = installedIntegrationsByName[ccrIntegration.package_name];
+      if (!localIntegrationSO) {
+        return {
+          ...ccrIntegration,
+          sync_status: 'failed' as SyncStatus.FAILED,
+          error: `Installation not found`,
+        };
+      }
+      if (ccrIntegration.package_version !== localIntegrationSO?.attributes.version) {
+        return {
+          ...ccrIntegration,
+          sync_status: 'failed' as SyncStatus.FAILED,
+          error: `Installed version: ${localIntegrationSO?.attributes.version}`,
+        };
+      }
+      if (localIntegrationSO?.attributes.install_status !== 'installed') {
+        return {
+          ...ccrIntegration,
+          sync_status: 'failed' as SyncStatus.FAILED,
+          error: `Installation status: ${localIntegrationSO?.attributes.install_status}`,
+        };
+      }
+      return {
+        ...ccrIntegration,
+        sync_status: 'completed' as SyncStatus.COMPLETED,
+        updated_at: localIntegrationSO?.updated_at,
+      };
+    }
+  );
+  return { integrations: integrationsStatus };
+};
+
+const compareCustomAssets = async (
+  esClient: ElasticsearchClient,
+  logger: Logger,
+  ccrCustomAssets: { [key: string]: CustomAssetsData }
+) => {
+  if (!ccrCustomAssets) return;
+
+  const abortController = new AbortController();
+
+  try {
+    const installedPipelines = await getPipeline(esClient, abortController);
+    const installedComponentTemplates = await getComponentTemplate(esClient, abortController);
```

Comment:
I went for the solution of fetching all the pipelines and component template and keep them on memory, instead of doing a call for each one in the loop below. @juliaElastic do you think it could become a performance issue?

Reply:
We should only compare pipelines and component templates that match `'*@custom'` only, like it's done in `custom_assets.ts`. The other assets are already installed when the package is installed.